libdnf/transaction/Swdb.cpp

```diff
@@ -55,22 +55,42 @@ Swdb::Swdb(const std::string &path)
   , autoClose(true)
 {
     if (path == ":memory:") {
-        // writing to an in-memory database
+        // connect to an in-memory database as requested
         conn = std::make_shared< SQLite3 >(path);
         Transformer::createDatabase(conn);
     } else if (!pathExists(path.c_str())) {
-        // writing to a file that doesn't exist and must be created
-
-        // extract persistdir from path - "/var/lib/dnf/"
-        auto found = path.find_last_of("/");
-
-        Transformer transformer(path.substr(0, found), path);
-        transformer.transform();
-
-        conn = std::make_shared< SQLite3 >(path);
+        if (geteuid() == 0) {
+            // database doesn't exist, running under root
+            // connect to a new database and migrate existing data into it
+
+            // extract persistdir from path - "/var/lib/dnf/"
+            auto found = path.find_last_of("/");
+            Transformer transformer(path.substr(0, found), path);
+            transformer.transform();
+            conn = std::make_shared< SQLite3 >(path);
+        } else {
+            try {
+                // database doesn't exist, running under unprivileged user
+                // connect to a new database and initialize it
+                conn = std::make_shared< SQLite3 >(path);
+                Transformer::createDatabase(conn);
+            } catch (std::exception &) {
+                // database initialization failed, connect to an in-memory database instead
+                conn = std::make_shared< SQLite3 >(":memory:");
+                Transformer::createDatabase(conn);
+            }
+        }
     } else {
-        // writing to an existing file
-        conn = std::make_shared< SQLite3 >(path);
+        try {
+            // connect to an existing database
+            conn = std::make_shared< SQLite3 >(path);
```

Comment:
Please:
conn = std::make_shared<SQLite3>(path);

Reply:
I turned it into ``std::make_shared<SQLite3>(path)``; I suppose GitHub ate the ``<SQLite>`` part in your comment.